packages/vitest/src/node/test-run.ts

```diff
@@ -0,0 +1,161 @@
+import type { HookOptions, TestCase, TestModule } from './reporters/reported-tasks'
+import type { TestSpecification } from './spec'
+
+export class TestRun {
+  private tests = emptyCounters()
+  private suites = emptyCounters()
+
+  constructor(private vitest: Vitest) {}
+
+  async start(specifications: TestSpecification[]) {
+    this.tests = emptyCounters()
+    this.suites = emptyCounters()
+    this.suites.total = specifications.length
+
+    await this.vitest.report('onTestRunStart', specifications)
+  }
+
+  async enqueued(project: TestProject, file: RunnerTestFile) {
+    this.vitest.state.collectFiles(project, [file])
+    const testModule = this.vitest.state.getReportedEntity(file) as TestModule
+    await this.vitest.report('onTestModuleQueued', testModule)
+  }
+
+  async collected(project: TestProject, files: RunnerTestFile[]) {
+    this.vitest.state.collectFiles(project, files)
+    await Promise.all([
+      this.vitest.report('onCollected', files),
+      ...files.map((file) => {
+        const testModule = this.vitest.state.getReportedEntity(file) as TestModule
+        return this.vitest.report('onTestModuleCollected', testModule)
+      }),
+    ])
+  }
+
+  async log(log: UserConsoleLog) {
+    this.vitest.state.updateUserLog(log)
+    await this.vitest.report('onUserConsoleLog', log)
+  }
+
+  async updated(update: TaskResultPack[]) {
+    this.vitest.state.updateTasks(update)
+
+    // These are used to guarantee correct reporting order
+    const runningTestModules: TestModule[] = []
+    const finishedTestModules: TestModule[] = []
+
+    const runningTestCases: TestCase[] = []
+    const finishedTestCases: TestCase[] = []
+
+    const startingHooks: HookOptions[] = []
+    const endingHooks: HookOptions[] = []
+
+    for (const [id,,,events] of update) {
+      const task = this.vitest.state.idMap.get(id)
+      const entity = task && this.vitest.state.getReportedEntity(task)
+
+      if (!entity) {
+        continue
+      }
+
+      for (const event of events) {
+        if (event === 'suite-prepare' && entity.type === 'module') {
+          runningTestModules.push(entity)
+        }
+
+        if (event === 'suite-finished' && entity.type === 'module') {
+          finishedTestModules.push(entity)
+
+          // Skipped tests need to be reported manually once test module has finished
+          for (const test of entity.children.allTests()) {
+            if (test.result().state === 'skipped') {
+              finishedTestCases.push(test)
+            }
+          }
+        }
+
+        if (event === 'test-prepare' && entity.type === 'test') {
+          runningTestCases.push(entity)
+        }
+
+        if (event === 'test-finished' && entity.type === 'test') {
+          finishedTestCases.push(entity)
+        }
+
+        if ((event === 'suite-hook-start' || event === 'suite-hook-end') && entity.task.result?.hooks) {
+          for (const hook of Object.keys(entity.task.result.hooks)) {
```

Comment:
Wouldn't it report all hooks that it has? For example, after `afterAll`, the task can have `beforeAll`, `beforeEach`, `afterEach` and `afterAll` in this object

Reply:
I haven't been able to get this part working fine yet. Feel free to change it, the test case in `test-run.test.ts` should print tasks correctly once fixed.